File: nacl/hello_world/hello_world.cc
#include "hello_world.h"

void HelloWorldInstance::
HandleMessage(const pp::Var& var_message) 
{
    if (!var_message.is_string()) {
        return;
    }
    std::string message = var_message.AsString();

    if (message == "hello") {
        pp::Var var_reply = pp::Var("hello from NaCl");
        PostMessage(var_reply);
    }
}

File: nacl/hello_world/hello_world.js
HelloWorldModule = null;  // Global application object.
statusText = 'NO-STATUS';

function moduleDidLoad() {
    HelloWorldModule = document.getElementById('hello_world');
    updateStatus('SUCCESS');

    HelloWorldModule.postMessage('hello');
}

function handleMessage(message_event) {
    alert(message_event.data);
}

function pageDidLoad() {
    if (HelloWorldModule == null) {
        updateStatus('LOADING...');
    } else {
        updateStatus();
    }
}

function updateStatus(opt_message) {
    if (opt_message)
        statusText = opt_message;
    var statusField = document.getElementById('status_field');
    if (statusField) {
        statusField.innerHTML = statusText;
    }
}
